### DISCUSSION TITLE
v1.4.0 

### DISCUSSION THREAD

**Discussion by boelnasr** [boelnasr/ManipulaPy#74] [Announcements]
## The unified compute backend system

The same kinematics, dynamics, planning and control code now runs on **NumPy, CuPy, PyTorch, or JAX** behind one dispatch API — with a tested differentiable contract on the core math under PyTorch and JAX.

This is an internal refactor. **The public API is unchanged** and the return contract under the default NumPy backend is frozen, so existing code and the ROS wrapper keep working untouched.

```python
from ManipulaPy.backend import use_backend

with use_backend("torch"):
    T = robot.forward_kinematics(theta)      # returns a torch tensor
    J = robot.jacobian(theta)                # gradients flow through
```

Gradients of forward kinematics and inverse dynamics come from `torch.autograd` / `jax.grad` directly — no finite differences, and no second implementation to keep in sync.

---

## Correctness fixes that affect every user

Building the gradient contract exposed **three real defects in the shipped SE(3)/SO(3) code**. All three produced wrong or non-finite results on **NumPy too** — not just on the new backends. If you use `MatrixLog6`, `MatrixLog3` or `logm`, upgrade for these alone.

- **`MatrixLog6` discarded small rotations.** Below θ = 1e-6 the entire rotational block was zeroed, so a genuine 5e-7 rotation came back as pure translation. The small-angle branch was unnecessary — pure translation falls out of the general formula.
- **Log gradients were NaN near the identity.** `jax.jacrev(MatrixLog3)` returned NaN for every θ below ~1e-2 while the values themselves looked correct.
- **The rotation angle was ill-conditioned near π.** `arccos(cos θ)` loses precision because `cos θ` is quadratic in the gap; now `atan2(|vee|/2, cos θ)`, which is linear in it.

Also fixed:

- **`package://` mesh resolution could escape the robot-description directory.** Candidates are now contained within the root that produced them, and symlinks pointing out of tree are refused.
- **Silent zero acceleration in forward-dynamics trajectories** — a `TypeError` on torch dtypes was swallowed by a broad `except`, zeroing acceleration at every step.
- **Packaging** — the sdist no longer sweeps developer tool configuration out of the working tree.

---

## Installing

```bash
pip install ManipulaPy                 # NumPy only — the default
pip install "ManipulaPy[pytorch]"      # + PyTorch
pip install "ManipulaPy[jax-cpu]"      # + JAX (CPU)
pip install "ManipulaPy[jax-cuda]"     # + JAX (CUDA 12, Linux)
pip install "ManipulaPy[cuda]"         # + CuPy (CUDA 12)
```

Backends register lazily behind an import probe, so a base install never imports torch or jax. Python 3.9–3.12 are supported; the **JAX extras require 3.10+**, since JAX itself dropped 3.9.

---

## Upgrade notes

- **Return types follow the active backend.** Under a non-default backend, public methods return that backend's native arrays by design. The frozen return contract applies to NumPy.
- **Non-ROS sibling directories are no longer auto-discovered by name** during `package://` resolution. Pin them explicitly with `add_package()` or `add_search_path()`.
- **Only `utils`, `kinematics`, `dynamics` and `singularity` carry a gradient guarantee.** Every other module runs on all four backends via host-boundary conversion, with no autodiff guarantee.
- **JAX eager dispatch is ~40× slower than NumPy on small problems.** Any wall-clock-budgeted algorithm behaves differently — notably `TracIKSolver`, which will not converge within its 200 ms default. The default is deliberately unchanged; widen `timeout` explicitly.
- **The JAX backend enables `jax_enable_x64` on import.** That is process-global JAX state and is not reverted.

---

## TPUs are not supported

There is no `jax-tpu` extra. The contract was run on real one-chip TPU v5e hardware and 9 of its 18 checks failed, for two reasons no tolerance change addresses:

- XLA:TPU implements neither float64 `LuDecomposition` nor int64 `dot`, so `inv`, `solve`, `mass_matrix` and `inverse_dynamics` raise `UNIMPLEMENTED`.
- The float64 matmuls that *do* execute are emulated on a bf16-native MXU and return float32-level accuracy — forward kinematics deviated from NumPy by ~3e-8 under `Precision.HIGHEST`, against a 1e-10 contract.

Because the JAX backend enables `jax_enable_x64` unconditionally, a TPU cannot meet its contract. Supporting one would require a per-platform precision domain, not a tolerance change.

---

## Verification

Every job green on `9ba7955`, the exact commit this release tags:

| axis | how it was proven |
| --- | --- |
| NumPy · Torch-CPU · JAX-CPU | hosted runners |
| **CuPy · Torch-CUDA · JAX-CUDA** | live CUDA hardware, gated behind a `gpu-axes-passed` marker tied to this SHA |
| optional deps (OpenCV, PyBullet, YOLO, scikit-learn) | real imports asserted, zero skips tolerated |
| regression | Python 3.9 · 3.10 · 3.11 · 3.12 |

Full suite: **1504 passed, 47 skipped**. Publishing to PyPI is gated on live-GPU evidence for this exact commit — a release cannot be cut from a SHA that has not run on real hardware.

Full details in [CHANGELOG.md](https://github.com/boelnasr/ManipulaPy/blob/main/CHANGELOG.md).

<hr /><em>This discussion was created from the release <a href='https://github.com/boelnasr/ManipulaPy/releases/tag/v1.4.0'>v1.4.0 </a>.</em>
